scheduler/scheduler.cpp

```diff
@@ -624,72 +626,104 @@ static string envs_match(CompileServer *cs, const Job *job)
-            assert(jobs.find((*it2)->id()) != jobs.end());
-        }
-    }
+            // Check if remote & if remote allowed
+            if (!cs->check_remote(job)) {
+                trace() << cs->nodeName() << " fails remote job check\n";
+                return false;
+            }
+        
+            return true;
+        });
+    return eligible;
+}
 
-    for (map<unsigned int, Job *>::const_iterator it = jobs.begin();
-            it != jobs.end(); ++it) {
-        Job *j = it->second;
+static CompileServer *pick_server_random(list<CompileServer *> &eligible)
+{
+    auto iter = eligible.cbegin();
+    std::advance(iter, random() % eligible.size());
+    return *iter;
+}
 
-        if (j->state() == Job::COMPILING) {
-            CompileServer *cs = j->server();
-            list<Job *> jobList = cs->jobList();
-            assert(find(jobList.begin(), jobList.end(), j) != jobList.end());
+static CompileServer *pick_server_least_busy(list<CompileServer *> &eligible) {
+    eligible.sort([] (const CompileServer *a, const CompileServer *b) {
+        return a->jobList().size() < b->jobList().size();
+    });
+    return *eligible.cbegin();
+}
+
+static CompileServer *pick_server_round_robin(Job *job, list<CompileServer *> &eligible)
+{
+    unsigned int oldest_job = 0;
+    CompileServer *selected = nullptr;
+
+    for (CompileServer * const cs: eligible) {
+        if (oldest_job < cs->lastPickedId() ||
+                (!cs->lastPickedId() && selected && !selected->lastPickedId() && !envs_match(cs, job).empty())) {
+            selected = cs;
+            oldest_job = cs->lastPickedId();
         }
     }
+    return selected;
+}
 
-#endif
+static CompileServer *pick_server_new(Job *job, list<CompileServer *> &eligible)
+{
+    CompileServer *selected = nullptr;
 
-    /* if the user wants to test/prefer one specific daemon, we look for that one first */
-    if (!job->preferredHost().empty()) {
-        for (CompileServer* const cs : css) {
-            if (cs->matches(job->preferredHost()) && cs->is_eligible_now(job)) {
-#if DEBUG_SCHEDULER > 1
-                trace() << "taking preferred " << cs->nodeName() << " " <<  server_speed(cs, job, true) << endl;
-#endif
-                return cs;
+    for (CompileServer * const cs: eligible) {
+        if ((cs->lastCompiledJobs().size() == 0) && (cs->jobList().size() == 0) && cs->maxJobs()) {
+            if (!selected) {
+                selected = cs;
+            } else if (!envs_match(cs, job).empty()) {
+                // if there is one server that already got the environment and one that
+                // hasn't compiled at all, pick the one with environment first
+                selected = cs;
             }
         }
-
-        return nullptr;
     }
+    return selected;
+}
 
+static CompileServer *pick_server_fastest(Job *job, list<CompileServer *> &eligible, uint8_t weight)
```

Comment:
While you are here, you should deprioritize any server that is more than half full. Maybe by reducing its assumed speed by 50 or even 75 percent. This is based on the assumption that essentially all modern systems use SMT/HyperThreads, so only half of your "cores" are real, and adding more threads past that, while still a marginal improvement, doesn't help quite as much as adding a real core.

A while back, I deployed this tweak[1] to a modified scheduler in the MongoDB NYC office and IIRC (it was many years ago...) got something like a 25% speedup on clean build times. Do we have any ability to test modified scheduler algorithms on our new cluster?

[1] I think I actually used a very strong discounting of `speed *= 0.1` to really try hard to fill up physical cores before touching logical ones, and also to get the secondary benefit of not consuming all cores on anyone's desktops unless the entire cluster was more than half full (which rarely happened at that point because we had >1000 cores, and most compiles used 2-300).

Reply:
The algorithms I'm adding are "better" in the sense because they assume all hosts are essentially equivalent but may be receiving different jobs that affect its perceived speed. This is the case for us and anyone else using hosts in the cloud, but it might not be true of in-house situations. I don't like this algorithm at all, but I'm trying to keep the defaults as similar to what they were before for now.

In all honesty, this is perhaps a large change for a point release, but I'm trying to keep it as small as I can (for now) while adding what we need. What you're suggesting is closer to a major version change - which I would like to see happen sooner rather than later, but I want something incremental first.

Reply:
It does seem like future optimizations of this algorithms could account for the number of CPUs that each server has, but that need not be part of this.  (and if made might be a new user selected able algorithm leaving these in place)

Reply:
@HenryMiller1 That's essentially what the "least busy" algorithm is trying to do, except with the maximum number of configured jobs substituting for the number of CPUs available. Here, I think "fastest" can and should mean what it always has - the host we expect will complete the task soonest, which isn't necessarily the least-busy one.